core/templates/services/question-validation.service.ts

```diff
@@ -53,6 +53,34 @@ export class QuestionValidationService {
     );
   }
 
+  hasDuplicateRules(question: Question): boolean {
+    const answerGroups = question._stateData.interaction.answerGroups;
+    const ruleSet = new Set<string>();
+
+    for (const group of answerGroups) {
+      for (const rule of group.rules) {
+        const ruleHash = `${rule.type}:${this.normalizeRuleInputs(rule.inputs)}`;
+
+        if (ruleSet.has(ruleHash)) {
+          return true;
+        }
+        ruleSet.add(ruleHash);
+      }
+    }
+
+    return false;
+  }
+
+  normalizeRuleInputs(inputs: Record<string, unknown>): string {
+    const sortedKeys = Object.keys(inputs).sort();
+    const normalizedInputs: Record<string, unknown> = {};
+
+    for (const key of sortedKeys) {
+      normalizedInputs[key] = inputs[key];
```

Comment:
Is the order in Record guaranteed to be fixed? I couldn't find documentation that states that, so it would be great if you could link to a reference (if you've validated this).

Reply:
The order of properties in a Record is not guaranteed, especially when dealing with objects that have non-integer keys.Is using a Map acceptable for this scenario? A Map guarantees insertion order, which seems to be what we need for normalizing the rule inputs.